Rust generator doesn't handle tuples of structs

```
error[E0308]: mismatched types
   --> ffi.rs:414:10
    |
414 |         (self.element1.to_rust(), self.element2.to_rust())
    |          ^^^^^^^^^^^^^^^^^^^^^^^ expected enum `ffi::CaseType`, found enum `api::CaseType`

```

--- a/tools/generator/filters/rust.py
+++ b/tools/generator/filters/rust.py
@@ -91,14 +91,14 @@ def rust_api_output_type(ctx, value: str, api_mod_path='') -> str:
 
     if is_array(value):
         return 'Vec<{}>'.format(
-            rust_api_output_type(ctx, get_array_inner(value))
+            rust_api_output_type(ctx, get_array_inner(value), api_mod_path)
         )
     elif value in basic_types:
         return basic_types[value]
     elif as_struct and is_tuple(as_struct):
         return '({})'.format(
             ', '.join(
-                rust_api_output_type(ctx, field)
+                rust_api_output_type(ctx, field, api_mod_path)
                 for _, field, _ in ctx['game'].get_struct(value)['str_field']
             ))
     else:
@@ -119,12 +119,14 @@ def rust_api_input_type(
             return '&[impl AsRef<str>]'
         else:
             return '&[impl Borrow<{}>]'.format(
-                rust_api_input_type(ctx, get_array_inner(value), skip_ref=True)
+                rust_api_input_type(
+                    ctx, get_array_inner(value), api_mod_path, True
+                )
             )
     elif as_struct and is_tuple(as_struct):
         return '({})'.format(
             ', '.join(
-                rust_api_input_type(ctx, field)
+                rust_api_input_type(ctx, field, api_mod_path)
                 for _, field, _ in ctx['game'].get_struct(value)['str_field']
             ))
     elif not rust_is_copy(ctx, value):

--- a/tools/generator/templates/player/rust/ffi.rs.jinja2
+++ b/tools/generator/templates/player/rust/ffi.rs.jinja2
@@ -4,7 +4,7 @@
 //! Types and conversions for the C interface
 //!
 //! Please use the tools defined in `api.rs` to interact with the API for
-//! {{ game.name }}.
+//! {{ game.name }}.
 
 #![allow(clippy::unit_arg)]
 #![allow(clippy::unused_unit)]
@@ -99,14 +99,14 @@ pub trait RustToC<T> {
 {% with ffi_type=type|rust_ffi_type, api_type=type|rust_api_output_type %}
 // Conversions for {{ type }}
 
-impl CToRust<{{ api_type }}> for {{ ffi_type }} {
-    unsafe fn to_rust(self) -> {{ api_type }} {
+impl CToRust<{{ api_type }}> for {{ ffi_type }} {
+    unsafe fn to_rust(self) -> {{ api_type }} {
         self
     }
 }
 
-impl RustToC<{{ ffi_type }}> for {{ api_type }} {
-    unsafe fn to_c(&self) -> {{ ffi_type }} {
+impl RustToC<{{ ffi_type }}> for {{ api_type }} {
+    unsafe fn to_c(&self) -> {{ ffi_type }} {
         *self
     }
 }
@@ -193,7 +193,7 @@ where
 {% with name=enum.enum_name|camel_case %}
 // Conversions for {{ enum.enum_name }}
 
-impl CToRust<api::{{ name }}> for {{ name }} {
+impl CToRust<api::{{ name }}> for {{ name }} {
     unsafe fn to_rust(self) -> api::{{ name }} {
         match self {
             {% for field_name, _ in enum.enum_field %}
@@ -203,8 +203,8 @@ impl CToRust<api::{{ name }}> for {{ name }} {
     }
 }
 
-impl RustToC<{{ name }}> for api::{{ name }} {
-    unsafe fn to_c(&self) -> {{ name }} {
+impl RustToC<{{ name }}> for api::{{ name }} {
+    unsafe fn to_c(&self) -> {{ name }} {
         match self {
             {% for field_name, _ in enum.enum_field %}
             api::{{ name }}::{{ field_name|camel_case }} => {{ name }}::{{ field_name|camel_case }},
@@ -221,13 +221,13 @@ impl RustToC<{{ name }}> for api::{{ name }} {
 // Conversions for {{ struct.str_name }}
 
 {% with api_type=struct.str_name|rust_api_output_type(api_mod_path='api::') %}
-impl CToRust<{{ api_type }}> for {{ ffi_type }} {
-    unsafe fn to_rust(self) -> {{ api_type }} {
+impl CToRust<{{ api_type }}> for {{ ffi_type }} {
+    unsafe fn to_rust(self) -> {{ api_type }} {
         {% if struct is tuple %}
         {% set separator = joiner(', ') %}
-        ({% for field, _, _ in struct.str_field %}{{ separator() }}self.{{ field }}.to_rust(){% endfor %})
+        ({% for field, _, _ in struct.str_field %}{{ separator() }}self.{{ field }}.to_rust(){% endfor %})
         {% else %}
-        {{ api_type }} {
+        {{ api_type }} {
             {% for field_name, _, _ in struct.str_field %}
             {{ field_name }}: self.{{ field_name }}.to_rust(),
             {% endfor %}
@@ -238,8 +238,8 @@ impl CToRust<{{ api_type }}> for {{ ffi_type }} {
 {% endwith %}
 
 {% with api_type=struct.str_name|rust_api_input_type(api_mod_path='api::', skip_ref=True) %}
-impl RustToC<{{ ffi_type }}> for {{ api_type }} {
-    unsafe fn to_c(&self) -> {{ ffi_type }} {
+impl RustToC<{{ ffi_type }}> for {{ api_type }} {
+    unsafe fn to_c(&self) -> {{ ffi_type }} {
         {% if struct is tuple %}
         let ({{ struct.str_field|map('first')|join(', ') }}) = self;
 
@@ -279,9 +279,9 @@ extern {
 #[no_mangle]
 unsafe extern "C" {{ func|rust_prototype(ffi=True) }} {
     {% for arg, type, _ in func.fct_arg %}
-    let {{ arg }} = {{ arg }}.to_rust();
+    let {{ arg }} = {{ arg }}.to_rust();
     {% endfor %}
     {% set separator = joiner(', ') %}
-    crate::{{ func.fct_name }}({{ separator() }}{% for field, _, _ in func.fct_arg %}{{ field }}{% endfor %}).to_c()
+    crate::{{ func.fct_name }}({{ separator() }}{% for field, _, _ in func.fct_arg %}{{ field }}{% endfor %}).to_c()
 }
 {% endfor %}

